### PR TITLE
Updated CMasternodeGovernance::VoteForTicket and CMasternodeGovernance::AddTicket pre-conditions

### PR DESCRIPTION
**Pull request by TimothyBanks**
### A. Improving governance voting logic

**Currently any active masternode can register governance payment ticket (CMasternodeGovernance::AddTicket), we need to add two more checks:**

check that MN is indeed enabled, by calling - CMasternode::IsEnabled() and check that it is “old enough”, this can be checked by comparing masternode’s broadcast message time with the current timethere is already method for that CMasternode::IsBroadcastedWithinold enough is 30 days (17280 blocks)

Same as above for Voting for ticket - CMasternodeGovernance::VoteForTicket

**B. Improving winner decision**

Governance ticket winners is decided if ticket gets 51% of Yes votes and 10% of masternode voted for it (CGovernanceTicket::IsWinner)

Currently 10% is calculated from the number of ALL masternodes in the network, this has to be changed to use only number of Enabled MNs

use CMasternodeMan::CountEnabled

### CODE REVIEW

File: src/mnode/mnode-governance.cpp [hunk 1]
```diff
@@ -121,12 +121,38 @@ CAmount CMasternodeGovernance::IncrementTicketPaidAmount(CAmount payment, CGover
...
+        return false;
+    }
+
+    if (!masterNode.IsEnabled()){
+        strErrorRet = "Only enabled Master Node can vote";
+        return false;
+    }
+
+    constexpr static auto thirty_days = int{30 * 24 * 60};
+    if (masterNode.IsBroadcastedWithin(thirty_days)){
```
Comment (TimothyBanks) [inline src/mnode/mnode-governance.cpp:143]: I'm still a bit unsure how to set up the test to get this condition to pass.

Comment (TimothyBanks) [inline src/mnode/mnode-governance.cpp:143]: I'm still a bit unsure how to set up the test to get this condition to pass.

File: src/mnode/mnode-governance.cpp [hunk 2]
```diff
@@ -121,12 +121,38 @@ CAmount CMasternodeGovernance::IncrementTicketPaidAmount(CAmount payment, CGover
     return aAmountPaid;
 }
 
+bool CMasternodeGovernance::CanVote(std::string& strErrorRet) const
```
Comment (akobrin1) [inline src/mnode/mnode-governance.cpp:124]: can make it noexcept...
IsActiveMasterNode,IsEnabled,IsBroadcastedWithin - noexcept
Just make also CMasternodeMan::Get const noexcept- it does not throw any exceptions

File: src/mnode/mnode-governance.cpp [hunk 3]
```diff
@@ -121,12 +121,38 @@ CAmount CMasternodeGovernance::IncrementTicketPaidAmount(CAmount payment, CGover
...
+        strErrorRet = "Failure retrieving Master Node";
+        return false;
+    }
+
+    if (!masterNode.IsEnabled()){
+        strErrorRet = "Only enabled Master Node can vote";
+        return false;
+    }
+
+    constexpr static auto thirty_days_in_seconds = int32_t{30 * 24 * 60 * 60};
```
Comment (akobrin1) [inline src/mnode/mnode-governance.cpp:142]: you could just declare constexpr static int32_t (or int as the type of parameter in IsBroadcastedWithin function)

File: src/mnode/mnode-governance.h [hunk 4]
```diff
@@ -200,4 +200,7 @@ class CMasternodeGovernance
     void Clear();
 
     void UpdatedBlockTip(const CBlockIndex *pindex);
-};
+
+private:
+    bool CanVote(std::string& strErrorRet) const;
```
Comment (akobrin1) [inline src/mnode/mnode-governance.h:205]: noexcept

Comment (akobrin1) [reply src/mnode/mnode-governance.h:205]: Please make it protected & virtual - this way it is easier to write tests. Will just mock it later on.

**Review by akobrin1 — CHANGES_REQUESTED** [review]
Will definitely need to rebase on master - I made some changes in mn-governance.cpp|h files.

**Review by a-ok123 — APPROVED** [review]

**Pull request closed without merging**